Support for reading GeoKeyDirectory

gschulze opened the pull request:
This PR implements support for reading the GeoKeyDirectory.

The structure is based on https://github.com/developmentseed/aiocogeo-rs/blob/main/src/geo_key_directory.rs, thanks [kylebarron](https://github.com/kylebarron) for the hint!

@weiji14 I split this off the coordinate transformation branch, as this is a prerequisite for debugging stuff conveniently in QGIS.

weiji14 commented on src/geo_key_directory.rs:
```diff
@@ -0,0 +1,647 @@
+use tiff::tags::Tag;
+use tiff::{TiffError, TiffFormatError, TiffResult};
+
+#[derive(Debug, PartialEq)]
+pub struct GeoKeyDirectory {
```
```suggestion
/// The GeoKeyDirectoryTag Requirements Class specifies the requirements for
/// implementing the reserved GeoKeyDirectoryTag TIFF tag.
///
/// Ref: https://docs.ogc.org/is/19-008r4/19-008r4.html#_requirements_class_geokeydirectorytag
#[derive(Debug, PartialEq)]
pub struct GeoKeyDirectory {
```

gschulze replied:
Done.

weiji14 commented on src/geo_key_directory.rs:
```diff
@@ -0,0 +1,647 @@
+    }
+}
+
+#[derive(Debug, TryFromPrimitive, IntoPrimitive)]
+#[repr(u16)]
+enum GeoKeyDirectoryTag {
```
```suggestion
// GeoTIFF key names and IDs
//
// Ref: https://docs.ogc.org/is/19-008r4/19-008r4.html#_summary_of_geokey_ids_and_names
#[derive(Debug, TryFromPrimitive, IntoPrimitive)]
#[repr(u16)]
enum GeoKeyDirectoryTag {
```

gschulze replied:
Done.

weiji14 commented on src/geo_key_directory.rs:
```diff
@@ -0,0 +1,647 @@
+    fn get_short(
+        key_tag: GeoKeyDirectoryTag,
+        location_tag: Option<Tag>,
+        count: u16,
+        offset: u16,
+    ) -> TiffResult<u16> {
+        if location_tag.is_some() {
+            return Err(TiffError::FormatError(TiffFormatError::Format(format!(
+                "Key `{key_tag:?}` did not have the expected value type."
+            ))));
+        }
+
+        if count != 1 {
```
Could use a ref here instead, and change all the `*count` to `count` above?

```suggestion
        count: &u16,
        offset: u16,
    ) -> TiffResult<u16> {
        if location_tag.is_some() {
            return Err(TiffError::FormatError(TiffFormatError::Format(format!(
                "Key `{key_tag:?}` did not have the expected value type."
            ))));
        }

        if count != &1 {
```

gschulze replied:
I can do that, but I'm unsure whether I completely understand the rationale. I think you want me to change it to a reference because of the contract of the equality operator: `fn eq(&self, &other: Self): bool`. Is there a performance impact to this, or is it just a matter of style? I can imagine the compiler will optimize this anyway. For me, `*count == 1` is more intuitive than `count == &1`, as in my mind, I'm comparing values rather than references. Changing from `count: u16` to `count: &u16` also consumes more memory on the stack, four times as much on 64-bit systems. Anyway, I'm happy to change it if you want me to; I just wanted to share my thoughts. Is there any best-practices document you follow that I can relate to?

kylebarron replied:
I'd love to hear rationale either way or see some resources on this, but for types that implement `Copy` I think it's simpler to not use a reference in the signature parameter

weiji14 replied:
Ah ok, I was approaching this more as a matter of style rather than performance. Still learning Rust on the go, and didn't realize that references to u16 would be stored as 64-bit (!!), so happy for you to stick with `*count == 1`.

gschulze replied:
I'm no expert on this topic either, but inspecting the code via [Compiler Explorer](https://godbolt.org) might reveal more insights.

weiji14 commented on src/geo_key_directory.rs:
```diff
@@ -0,0 +1,647 @@
+        data: &[f64],
+        key_tag: GeoKeyDirectoryTag,
+        location_tag: Option<Tag>,
+        count: u16,
+        offset: u16,
+    ) -> TiffResult<f64> {
+        if location_tag != Some(Tag::GeoDoubleParamsTag) {
+            return Err(TiffError::FormatError(TiffFormatError::Format(format!(
+                "Key `{key_tag:?}` did not have the expected value type."
+            ))));
+        }
+
+        if count != 1 {
```
```suggestion
        count: &u16,
        offset: u16,
    ) -> TiffResult<f64> {
        if location_tag != Some(Tag::GeoDoubleParamsTag) {
            return Err(TiffError::FormatError(TiffFormatError::Format(format!(
                "Key `{key_tag:?}` did not have the expected value type."
            ))));
        }

        if count != &1 {
```

gschulze replied:
Done.

weiji14 commented on src/geo_key_directory.rs:
```diff
@@ -0,0 +1,647 @@
+        Self {
+            key_directory_version: 1,
+            key_revision: 1,
+            minor_revision: 1,
```
Assumes GeoTIFF version 1.1 as default, should be ok.

See requirements 2.9 under https://docs.ogc.org/is/19-008r4/19-008r4.html#_requirements_class_geokeydirectorytag:

> 
> http://www.opengis.net/spec/GeoTIFF/1.1/req/GeoKeyDirectoryTag.minorRevisionValue
> The MinorRevision for this standard SHALL be 0 or 1.
> 
> 0 = GeoTIFF 1.0 version. Accepted for existing files (GeoTIFF 1.0) that do not use values / codes that are suppressed in this version
> 
> 1 = this version 1.1. Recommended for production / writing a GeoTIFF file according to this standard (and developing software). Shall be used for files keys and codes not covered by GeoTIFF 1.0 (e.g. VerticalGeoKey).

gschulze replied:
I will put that in the doc-Comment of `GeoKeyDirectory`.

weiji14 commented on src/geo_key_directory.rs:
```diff
@@ -0,0 +1,647 @@
+        count: u16,
+        offset: u16,
+    ) -> TiffResult<u16> {
+        if location_tag.is_some() {
+            return Err(TiffError::FormatError(TiffFormatError::Format(format!(
+                "Key `{key_tag:?}` did not have the expected value type."
+            ))));
+        }
```
This check and error message is a little confusing to me at first glance. Would it be clearer if the match was something more explicit like `if location_tag is Tag::GeoDoubleParamsTag or Tag::GeoAsciiParamsTag`?

gschulze replied:
Agree.

weiji14 replied:
Nice, thanks for the explanation. If we could add an inline-comment to say that this if-check is to ensure that `TIFFTagLocation == 0`, that would be great.

weiji14 reviewed (commented):
Thanks @gschulze! Just some docstring and minor code suggestions/questions for now.

gschulze commented:
@weiji14 Thanks for your comments; I updated the PR and incorporated most of your suggestions. See my comment on the usage of `*count == 1` vs `count == &1` above, though.

weiji14 commented on src/geo_key_directory.rs:
```diff
@@ -0,0 +1,658 @@
+        count: u16,
+        offset: u16,
+    ) -> TiffResult<u16> {
+        if location_tag.is_some() {
```
Suggestion based on https://github.com/georust/geotiff/pull/19#discussion_r1769714908

```suggestion
        // Check that TIFFTagLocation == 0 so value is of SHORT type
        if location_tag.is_some() {
```

weiji14 reviewed (approved):
Thanks again @gschulze, just one small doc suggestion, otherwise feel free to merge!
